File: node/packages/middlewares/src/jwt.ts
import { Request, Response, NextFunction } from "express";
import jwt, { JwtHeader } from "jsonwebtoken";
import jwksClient, { JwksClient } from "jwks-rsa";
import { JwtClaims } from "./types";

const clients: Record<string, JwksClient> = {};

/**
 * Get or create a JWKS client for a given realm
 */
function getJwksClient(realm: string): JwksClient {
  if (!clients[realm]) {
    clients[realm] = jwksClient({
      jwksUri: `${process.env.KEYCLOAK_URL}/realms/${realm}/protocol/openid-connect/certs`,
    });
  }
  return clients[realm];
}

/**
 * Create a key resolver for a given realm
 */
function getKey(realm: string) {
  return (header: JwtHeader, callback: (err: Error | null, key?: string) => void) => {
    const client = getJwksClient(realm);
    if (!header.kid) return callback(new Error("No kid found in token header"));

    client.getSigningKey(header.kid, (err, key) => {
      if (err) return callback(err);
      const signingKey = key?.getPublicKey();
      callback(null, signingKey);
    });
  };
}

/**
 * Extract realm name from the token's `iss` claim
 */
function extractRealmFromToken(token: string): string | null {
  const decoded = jwt.decode(token) as { [key: string]: any } | null;
  const iss: string | undefined = decoded?.iss;
  if (!iss) return null;

  const match = iss.match(/\/realms\/([^/]+)$/);
  return match ? match[1] : null;
}

/**
 * Middleware: Verify JWT signature + claims using realm-specific JWKS
 */
export function verifyAndGetClaims(req: Request, res: Response, next: NextFunction) {
  const authHeader = req.headers["authorization"];
  if (!authHeader) return res.status(401).json({ error: "No Authorization header" });

  const token = authHeader.split(" ")[1];
  if (!token) return res.status(401).json({ error: "Invalid Authorization header format" });

  const realm = extractRealmFromToken(token);
  if (!realm) return res.status(400).json({ error: "Could not extract realm from token" });

  const issuer = `${process.env.KEYCLOAK_URL}/realms/${realm}`;
  const audience = "account";

  console.log(issuer, getKey(realm))

  jwt.verify(token, getKey(realm), {
    algorithms: ["RS256"],
    issuer, audience
  }, (err, decoded) => {
    if (err) {
      return res.status(401).json({ error: "Invalid token", details: err.message });
    }
    req.claims = decoded as JwtClaims;
    next();
  });
}

/**
 * Middleware: Just decode the JWT without verifying (trust external proxy did verification)
 */
export function getClaims(req: Request, res: Response, next: NextFunction) {
  const authHeader = req.headers["authorization"];
  if (!authHeader) return res.status(401).json({ error: "No Authorization header" });

  const token = authHeader.split(" ")[1];
  if (!token) return res.status(401).json({ error: "Invalid Authorization header format" });

  try {
    const decoded = jwt.decode(token);
    if (!decoded) return res.status(400).json({ error: "Failed to decode JWT" });

    req.claims = decoded as JwtClaims;
    next();
  } catch (err) {
    return res.status(400).json({
      error: "Error decoding token",
      details: (err as Error).message,
    });
  }
}

File: python/packages/middlewares/src/package_middlewares/jwt.py
import os
import re
from typing import Dict, Optional
from urllib.request import urlopen

import jwt
from fastapi import HTTPException, Request, status
from jwt import PyJWKClient

from .types import JwtClaims

JWKS_CLIENTS: Dict[str, PyJWKClient] = {}

def get_jwks_client(realm: str) -> PyJWKClient:
    if realm not in JWKS_CLIENTS:
        keycloak_url = os.getenv("KEYCLOAK_URL")
        if not keycloak_url:
            raise ValueError("KEYCLOAK_URL environment variable not set")
        jwks_uri = f"{keycloak_url}/realms/{realm}/protocol/openid-connect/certs"
        JWKS_CLIENTS[realm] = PyJWKClient(jwks_uri)
    return JWKS_CLIENTS[realm]

async def get_key(realm: str, kid: str) -> str:
    client = get_jwks_client(realm)
    signing_key = client.get_signing_key(kid)
    return signing_key.key

def extract_realm_from_token(token: str) -> Optional[str]:
    decoded = jwt.decode(token, options={"verify_signature": False})
    iss = decoded.get("iss")
    if not iss:
        return None
    match = re.search(r"/realms/([^/]+)$", iss)
    return match.group(1) if match else None

async def verify_and_get_claims(request: Request):
    auth_header = request.headers.get("Authorization")
    if not auth_header:
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="No Authorization header")

    token = auth_header.split(" ")[1]
    if not token:
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Invalid Authorization header format")

    realm = extract_realm_from_token(token)
    if not realm:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Could not extract realm from token")

    try:
        unverified_header = jwt.get_unverified_header(token)
        kid = unverified_header["kid"]
        key = await get_key(realm, kid)
        decoded_token = jwt.decode(token, key=key, algorithms=["RS256"], audience="account", issuer=f"{os.getenv("KEYCLOAK_URL")}/realms/{realm}")
        request.state.claims = JwtClaims(**decoded_token)
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail=f"Invalid token: {e}")

async def get_claims(request: Request):
    auth_header = request.headers.get("Authorization")
    if not auth_header:
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="No Authorization header")

    token = auth_header.split(" ")[1]
    if not token:
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Invalid Authorization header format")

    try:
        decoded_token = jwt.decode(token, options={"verify_signature": False})
        request.state.claims = JwtClaims(**decoded_token)
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=f"Error decoding token: {e}")
